### ISSUE TITLE
User stories

### ISSUE THREAD

**Issue by oliverbarnes** [decidim/docker#57]
@andreslucena suggested we outline the user stories for docker with Decidim that we're trying to support here, so it's clear what we're trying to accomplish with the builds and images.

Here are a some types of users and the stories we have gleaned so far. This will be updated as the discussion progresses.

## User types

These can overlap, and several times the same person might change hats and act as different user types:

### Instance (or App) Implementor
A developer working on a Decidim instance for a city, organization or another entity.

Has at least some knowledge about using the *nix command line and of how services work in GNU/Linux systems. Doesn't necessarily have knowledge about Ruby nor Ruby on Rails.

Probably doesn't know much about Docker.

### Instance (or App) Deployer

Either a developer or ops person (or both), with some experience deploying webapps in the infrastructure they're using.

May or may not have knowledge of Docker, but has enough ops skills to learn basics of deploying it on their environment.

### Module Developer
A developer working on a module extending or customizing Decidim. 

This might be in the context of implementing an instance or as part of Decidim framework work, but the workflow for developing the module is the same.

Knows the *nix command line, and has good knowledge of Ruby, Rails and Rails engines.

Might or not have Docker knowledge.

### Core Developer
Working on Decidim core and core modules.

Knows the *nix command line, and has advanced knowledge of Ruby, Rails and Rails engines.

Might or not have Docker knowledge.

## Stories
- **Instance Implementor** runs Decidim instance locally for the first time, to test-drive it. aka "Hello world" (#67)
- **Instance Implementor** clones/forks an existing Decidim instance repo (like decidim/demo.decidim.org) and runs it locally
- **Instance Implementor** clones/forks an existing Decidim instance repo (like decidim/demo.decidim.org) and deploys it
- **Instance Implementor** deploys a pre-configured Decidim instance image
- **Instance Implementor** generates a Decidim instance to configure and customize
- **Instance Implementor** dockerizes their instance for deployment and distribution
- **Module Developer** sets up their environment for module development
- **Core Developer** sets up their environment for core framework development

**Comment by andreslucena** [comment]
I would also add some technical knowledge of each one: 

* **Instance implementor**: has some knowledge about using the command line and how services work in GNU/Linux systems. Hasn't knowledge about Ruby nor Ruby on Rails. 
* **Module and Core Developers**: have knowledge about the command line, GNU/Linux systems, Ruby, and Ruby on Rails. 

I'm not so sure about whether they should have (or not) Docker knowledge.

**Comment by oliverbarnes** [comment]
Thanks, updated.

**Comment by andreslucena** [comment]
A proposal on one User Story to solve: 

----

As a Decidim implementer, I want to be able to put Decidim in production without knowing anything about the technology is built upon: Ruby on Rails, Docker, etc. For this I want to have an installation workflow similar to the one that [Discourse](https://github.com/discourse/discourse/blob/master/docs/INSTALL.md) has, where a person without knowledge of Rails can install it. The only knowledge that I need to have is using the command line. 

For this the only dependencies that I'll need to have are:
* Docker
* docker-compose

For instance for working in the demo app of Decidim you'll only need to run these commands:

```bash
git clone https://github.com/decidim/demo.decidim.org
cd demo.decidim.org
cp .env.example .env
docker-compose up -d
docker-compose run --rm exec app rails db:seed
```

All the configuration will need to be done through Environment Variables (see https://github.com/decidim/decidim/discussions/6918), so it's easier to configure everything.

Some restrictions: 
* All the technology should be FLOSS by default (ie not using Heroku/AWS or other proprietary services)
* Base Operating System is Ubuntu 20.04 LTS

**Comment by oliverbarnes** [comment]
Ok, based on the above and latest discussion over a call with @andreslucena, I've updated the user stories description with the following:

**New user type** 

Instance (or App) Deployer. Somebody with minimal web app deployment experience in a given environment the instance will be deployed to. Specially so with dedicated servers,  even if we maintain sophisticated deployment scripts.

** New user stories **

- Instance Implementor runs Decidim instance locally for the first time, to test-drive it. aka "Hello world"
- Instance Implementor clones/forks an existing Decidim instance repo (like decidim/demo.decidim.org) and runs it locally
- Instance Implementor clones/forks an existing Decidim instance repo (like decidim/demo.decidim.org) and deploys it

I've also removed mentions of specific deployment environments. The deployment image should be agnostic and just depend on environment variables to be run, same on a dedicated server or on the cloud.
